src/hypothesis/internal/charstree.py

```diff
@@ -0,0 +1,310 @@
+
+__all__ = (
+    'ascii_tree',
+    'unicode_tree',
+    'categories',
+    'category_by_codepoint',
+    'codepoints',
+    'codepoints_for_category',
+    'filter_tree',
+    'random_codepoint',
+)
+
+
+ASCII_TREE = {}
+UNICODE_TREE = {}
+
+
+def ascii_tree():
+    """Returns tree for ASCII characters."""
+    global ASCII_TREE
+    if not ASCII_TREE:
+        ASCII_TREE = filter_tree(unicode_tree(), max_codepoint=127)
+    return ASCII_TREE
+
+
+def unicode_tree():
+    """Returns tree of Unicode characters."""
+    global UNICODE_TREE
+    if not UNICODE_TREE:
+        try:
+            UNICODE_TREE = load_tree()
+        except:
+            UNICODE_TREE = make_tree()
+            try:
+                dump_tree(UNICODE_TREE)
+            except:  # pragma: no cover
+                pass
+    return UNICODE_TREE
+
+
+def categories(tree):
+    """Returns list of all categories in specified tree."""
+    return list(tree)
+
+
+def category_by_codepoint(tree, codepoint):
+    """Returns category by specified code point in the tree."""
+    categories = list(filter_tree(
+        tree, min_codepoint=codepoint, max_codepoint=codepoint))
+    if not categories:
+        return None
+    assert len(categories) == 1
+    return categories[0]
+
+
+def codepoints(tree):
+    """Iterates over all code points in specified tree.
+
+    Code points get yielded in single stream category by category.
+
+    """
+    by_category = functools.partial(codepoints_for_category, tree)
+    return itertools.chain.from_iterable(map(by_category, categories(tree)))
+
+
+def codepoints_for_category(tree, category):
+    """Iterates over all code points of the specified category in the tree."""
+    for value in iter_values(tree[category]):
+        for cp in range(value[0], value[1] + 1):
+            yield cp
+
+
+def random_codepoint(tree, category, random):
+    """Returns random code point for specified category in the tree."""
+    base, values = select_values(
+        tree, category, lambda ns: random.choice(list(ns)))
+    assert values, 'no values found'
+    value = random.choice(values)
+    if value[0] == value[1]:
+        return base + value[0]
+    return random.randint(base + value[0], base + value[1])
+
+
+def filter_tree(tree, whitelist_categories=None, blacklist_categories=None,
+                blacklist_characters=None, min_codepoint=0,
+                max_codepoint=sys.maxunicode):
+    """Filters tree by Unicode categories and code points range."""
+    if whitelist_categories and blacklist_categories:
+        raise InvalidArgument('cannot have both white and black list of'
+                              ' categories at the same time')
+
+    if min_codepoint > max_codepoint:
+        raise InvalidArgument('min code point is greater than max')
+
+    categories = set(tree)
+    if whitelist_categories:
+        categories &= set(whitelist_categories)
+    if blacklist_categories:
+        categories ^= set(blacklist_categories)
```

Comment:
I don't think this is the right operator. Do you mean -=? This will add in any cartegories that are blacklisted but didn't pass the whitelist.